man/man1/drainage2cdf.1
.TH drainage2cdf 1  "USER COMMANDS"
.SH NAME
drainage2cdf \- convert CLM drainage file into NetCDF file
.SH SYNOPSIS
.B drainage2cdf
[\-var name] [\-compress level] [\-netcdf4]
.I soilfile.nc gridfile drainagefile netcdffile
.SH DESCRIPTION
Program converts CLM drainage data into NetCDF input data using the soil code NetCDF file. If drainage CLM file has 2 bands, then the second band is the length of the river and written as variable "riverlen" into the NetCDF file.
.SH OPTIONS
.TP
\-var name
Name of soil code variable in the soil code NetCDF file.
.TP
\-compress level
Compression level for NetCDF4 files. Valid levels are 0 (no compression) to 9.
.TP
\-netcdf4
File written is in NetCDF4 format, compress flag is enabled.
.TP
.I soilfile.nc
Soil code file in NetCDF format.
.TP
.I gridfile
Filename of grid data file.
.TP
.I drainagefile
Filename of CLM drainage data file.
.TP
.I netcdffile
Filename of NetCDF file created.
.SH EXAMPLES
.TP
Create drainage data:
.B drainage2cdf
-var soilcode soil.nc grid.clm drainage.clm drainage.nc
.PP
.SH EXIT STATUS
.B drainage2cdf
returns a zero exit status if the NetCDF file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
bin2cdf(1), country2cdf(1), reservoir2cdf(1), regridlpj(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Hester Biemans (hester.biemans@wur.nl)
- code review: Susanne Rolinski (rolinski@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `input_netcdf.cjson` changed to the default dataset of `input.cjson` but in NetCDF format.
- Code changed to compile under Windows OS.
- New function `getsprintf()` added to allocate and print formatted output into string. Function replaces call to `snprintf()` function.
- Default settings for `GIT_HASH` and `GIT_REPO` added.
- Duplicate filenames removed from list of inpute/output filenames in utility `lpjfiles`.

### Added

- Option `-int` added to utility `cdf2clm`.
- Option `-latlon` added to utility `cdf2coord` in order to change the order of the CLM grid file.
- Utility `cdf2reservoir` added to convert NetCDF reservoir file into CLM file.
- Utility `reservoir2cdf` added to convert CLM reservoir file into a NetCDF file using the soil code NetCDF file.
- Macro `NETCDF_INPUT` added in `lpjml_config.cjson` to enable NetCDF input.
- Reservoir, irrigation neighbor and drainage data can now be in NetCDF format:
```java
"drainage" :           { "fmt" : "cdf", "var" : "index", "name" : "cru_netcdf/drainage.nc"},
"neighb_irrig" :       { "fmt" : "cdf", "var" : "index", "name" : "cru_netcdf/neighb_irrig.nc"},
"river" :              { "fmt" : "cdf", "var" : "riverlen", "name" : "cru_netcdf/drainage.nc"},
"reservoir" :          { "fmt" : "cdf", "var" : "year", "name" : "cru_netcdf/reservoir.nc"},
"capacity_reservoir" : { "fmt" : "cdf", "var" : "capacity", "name" : "cru_netcdf/reservoir.nc"},
"area_reservoir" :     { "fmt" : "cdf", "var" : "area", "name" : "cru_netcdf/reservoir.nc"},
"inst_cap_reservoir" : { "fmt" : "cdf", "var" : "inst_cap", "name" : "cru_netcdf/reservoir.nc"},
"height_reservoir" :   { "fmt" : "cdf", "var" : "height", "name" : "cru_netcdf/reservoir.nc"},
"purpose_reservoir" :  { "fmt" : "cdf", "var" : "purpose", "name" : "cru_netcdf/reservoir.nc"},
```

### Fixed

- River length is now correctly read in `initdrain.c` for data in NetCDF format.
- Man page of `cvrtclm` corrected.
- Missing check for open NetCDF file added to `openclimate.c`.


## [5.9.18] - 2025-01-31

### Contributors

- author: Christoph Müller (cmueller@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Fixed

- `basetemp` and `hlimit` parameters fixed to fit to standard growing season inputs (issue #346).
- `laimax` values corrected from 5 to 7 for `temperate cereals`, `rice`, and `tropical cereals`


## [5.9.17] - 2025-01-29

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Sebastion Ostberg (ostberg@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)
- code review: Marie Hemmen (hemmen@pik-potsdam.de), Jens Heinke (heinke@pik-potsdam.de)

### Added

- Option `-cmd` added to `lpjsubmit` in order to execute a command before lpjml is called.
- Command `module list` added to slurm script to show all loaded modules.

### Removed

- Obsolete `#ifdef WITHOUT_NITROGEN` check removed from `pft.cjson`.

### Fixed

- `ALLOM3` parameter corrected to 4.0 for oil palm in `pft.cjson` (issue #369).
- Missing `break` added in `convert_water.c`.


## [5.9.16] - 2024-11-15

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- GIT repository and GIT hash printed in the LPJmL banner and in output of NetCDF and JSON files.
- Options `-repo` and `-hash` added to `lpjml` to print GIT repository and hash.
- The JSON file created by `createconfig` contains now name and hash of GIT repository.
- Command `lpjml -v` prints GIT repository and hash.
- Target `hash` added to Makefile to update hash after committing changes to the GIT repository.

### Changed

- `LPJ_VERSION` macro in `lpj.h` replaced by function `getversion()`.

### Removed

- Obsolete header columns cereal and maize removed from `fprintcountrypar.c`.

### Fixed

- Missing dependency on `gebuild.c` for target `lpjml` added in `src/Makefile`.
- `xiar` replaced by `ar` in `Makefile.hpc2024` in order to compile with `intel/oneAPI/2025.0.0`.
- Typos in error messages in `filesexist.c` and `cdf2soil.c`fixed.
- Missing check for open file added to `joingrid.c`.


## [5.9.15] - 2024-11-07

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- Option `-scale` added to `bin2cdf` in order to scale output.
- Option `-notime` added to `bin2cdf` in order to omit time axis.
- `time_bnds`, `lat_bnds`, and `lon_bnds` arrays added to NetCDF output for lpjml runs and the `bin2cdf` utility.
- Check for identical output filenames added in `fscanoutput.c`.

### Changed

- If `"with_days"` is set to true in the lpjml configuration file, time axis of yearly output is also set in units of days instead of years.
- Time in NetCDF output is always set in the middle between the lower and upper time boundaries.

### Removed

- Unused file `cflux_sum.c` removed.
- Conversion of obsolete file `manage.par` removed from utility `manage2js`.

### Fixed


src/utils/drainage2cdf.c
/**************************************************************************************/
/**                                                                                \n**/
/**                d  r  a  i  n  a  g  e  2  c  d  f  .  c                        \n**/
/**                                                                                \n**/
/**     Converts drainage clm file into NetCDF data file                           \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"
#ifdef USE_NETCDF
#include <netcdf.h>
#include <time.h>
#endif
#define error(rc) if(rc) {fprintf(stderr,"ERROR427: Cannot write '%s': %s.\n",argv[iarg+3],nc_strerror(rc)); nc_close(ncid);return EXIT_FAILURE;}

#define USAGE "Usage: %s [-var name] [-compress level] [-netcdf4] soilcode.nc grid.clm drainage.clm drainage.nc\n"
#define INDEX_LONG_NAME "index of cell into/from which water flows (ilat*nlon+ilon)"

int main(int argc,char **argv)
{
#ifdef USE_NETCDF
  char *var;
  var=NULL;
  const double *lon,*lat;
  int *index;
  Coord *grid_soil,*grid;
  FILE *file;
  Coordfile gridfile;
  Coord_netcdf soil;
  Coord resolution;
  Filename name;
  Header header;
  Bool swap,isnetcdf4;
  int iarg,i,n,nlon,nlat,j,offset[2]={0,0};
  int ncid,rc,lat_dim_id,lon_dim_id,lon_var_id,lat_var_id;
  int dim[2];
  int index_varid,len_varid;
  int miss=MISSING_VALUE_INT;
  int data[2];
  int version,compress;
  int src_cell,dst_cell;
  int *out;
  int s_len;
  char *s,*endptr;
  time_t t;
  char *cmdline;
  float *len=NULL,fmiss,glon,glat;
  unsigned int soilcode;
  String headername,line;
  compress=0;
  isnetcdf4=FALSE;
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-var"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-var'.\n"
                 USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        var=argv[++iarg];
      }
      else if(!strcmp(argv[iarg],"-netcdf4"))
        isnetcdf4=TRUE;
      else if(!strcmp(argv[iarg],"-compress"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Error: Missing argument after option '-compress'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        compress=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-compress'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
        if(compress<0 || compress>9)
        {
          fprintf(stderr,"Error: Invalid compression value %d, must be in [0,9].\n",
                  compress);
          return EXIT_FAILURE;
        }
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                USAGE,argv[iarg],argv[0]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<iarg+4)
  {
    fprintf(stderr,"Error: Missing argument(s).\n"
            USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  soil=opencoord_netcdf(argv[iarg],var,TRUE);
  if(soil==NULL)
    return EXIT_FAILURE;
  getresolution_netcdf(soil,&resolution);
  n=numcoord_netcdf(soil);
  if(n==-1)
  {
    fputs("ERROR165: Cannot get number of cells from soil code file.\n",stderr);
    return EXIT_FAILURE;
  }
  printf("%d cells found in `%s`.\n",n,argv[iarg]);
  grid_soil=newvec(Coord,n);
  check(grid_soil);
  index=newvec(int,n);
  check(index);
  lon=getlon_netcdf(soil,&nlon);
  lat=getlat_netcdf(soil,&nlat);
  for(i=0;i<n;i++)
  {
    if(readcoord_netcdf(soil,grid_soil+i,&resolution,&soilcode))
    {
      fprintf(stderr,"ERROR190: Cannot read coordinate from '%s' for cell %d.\n",
              argv[iarg],i);
      return EXIT_FAILURE;
    }
    for(j=0;j<nlon;j++)
      if(grid_soil[i].lon==lon[j])
      {
        offset[1]=j;
        break;
      }
    for(j=0;j<nlat;j++)
      if(grid_soil[i].lat==lat[j])
      {
        offset[0]=j;
        break;
      }
    index[i]=offset[0]*nlon+offset[1];
#ifdef DEBUG
    printf("%d %d\n",i,index[i]);
#endif
  }
  name.name=argv[iarg+1];
  name.fmt=CLM;
  gridfile=opencoord(&name,TRUE);
  if(gridfile==NULL)
    return EXIT_FAILURE;
  if(numcoord(gridfile)!=n)
  {
    fprintf(stderr,"Number of cells=%d in '%s' differs from %d in '%s'.\n",
            numcoord(gridfile),argv[iarg+1],n,argv[iarg]);
    return EXIT_FAILURE;
  }
  getcellsizecoord(&glon,&glat,gridfile);
  if(resolution.lon!=glon || resolution.lat!=glat)
  {
    fprintf(stderr,"Cell size (%g,%g) in '%s' differs from (%g,%g) in '%s'.\n",
            glon,glat,argv[iarg+1],resolution.lon,resolution.lat,argv[iarg]);
    return EXIT_FAILURE;
  }
  grid=newvec(Coord,n);
  check(grid);
  for(i=0;i<n;i++)
  {
    if(readcoord(gridfile,grid+i,&resolution))
    {
      fprintf(stderr,"Error reading cell %d in `%s`.\n",
              i,argv[iarg+1]);
      return EXIT_FAILURE;
    }
  }
  closecoord(gridfile);
  file=fopen(argv[iarg+2],"rb");
  if(file==NULL)
  {
    printfopenerr(argv[iarg+2]);
    return EXIT_FAILURE;
  }
  version=READ_VERSION;
  if(freadanyheader(file,&header,&swap,headername,&version,TRUE))
  {
    fprintf(stderr,"Invalid header in '%s'.\n",argv[iarg+2]);
    return EXIT_FAILURE;
  }
  if(resolution.lon!=header.cellsize_lon || resolution.lat!=header.cellsize_lat)
  {
    fprintf(stderr,"Cell size (%g,%g) in '%s' differs from (%g,%g) in '%s'.\n",
            header.cellsize_lon,header.cellsize_lat,argv[iarg+2],resolution.lon,resolution.lat,argv[iarg]);
    return EXIT_FAILURE;
  }
  if(header.ncell!=n)
  {
    fprintf(stderr,"Number of cells=%d in '%s' differs from %d in '%s'.\n",
            header.ncell,argv[iarg+2],n,argv[iarg]);
    return EXIT_FAILURE;
  }
  if(header.nbands>2)
  {
    fprintf(stderr,"Number of bands=%d in '%s' must be 1 or 2.\n",
            header.nbands,argv[iarg+2]);
    return EXIT_FAILURE;
  }
  if(version>2 && header.datatype!=LPJ_INT)
  {
    fprintf(stderr,"Datatype of file '%s' is %s, must be int.\n",argv[iarg+2],typenames[header.datatype]);
    return EXIT_FAILURE;
  }
  if(isnetcdf4)
    rc=nc_create(argv[iarg+3],(compress) ? NC_CLOBBER|NC_NETCDF4 : NC_CLOBBER,&ncid);
  else
    rc=nc_create(argv[iarg+3],NC_CLOBBER,&ncid);
  if(rc)
  {
    fprintf(stderr,"ERROR426: Cannot create file '%s': %s.\n",
            argv[iarg+3],nc_strerror(rc));
    return EXIT_FAILURE;
  }
  time(&t);
  cmdline=catstrvec(argv,argc);
  s_len=snprintf(NULL,0,"%s: %s",strdate(&t),cmdline);
  s=malloc(s_len+1);
  sprintf(s,"%s: %s",strdate(&t),cmdline);
  rc=nc_put_att_text(ncid,NC_GLOBAL,"history",strlen(s),s);
  error(rc);
  free(s);
  free(cmdline);
  rc=nc_def_dim(ncid,LAT_DIM_NAME,nlat,&lat_dim_id);
  error(rc);
  rc=nc_def_dim(ncid,LON_DIM_NAME,nlon,&lon_dim_id);
  error(rc);
  rc=nc_def_var(ncid,LAT_NAME,NC_DOUBLE,1,&lat_dim_id,&lat_var_id);
  error(rc);
  rc=nc_def_var(ncid,LON_NAME,NC_DOUBLE,1,&lon_dim_id,&lon_var_id);
  error(rc);
  rc=nc_put_att_text(ncid,lon_var_id,"units",strlen("degrees_east"),
                     "degrees_east");
  error(rc);
  rc=nc_put_att_text(ncid, lon_var_id,"long_name",strlen(LON_LONG_NAME),LON_LONG_NAME);
  error(rc);
  rc=nc_put_att_text(ncid, lon_var_id,"standard_name",strlen(LON_STANDARD_NAME),LON_STANDARD_NAME);
  error(rc);
  rc=nc_put_att_text(ncid, lon_var_id,"axis",strlen("X"),"X");
  error(rc);
  rc=nc_put_att_text(ncid,lat_var_id,"units",strlen("degrees_north"),
                     "degrees_north");
  error(rc);
  rc=nc_put_att_text(ncid, lat_var_id,"long_name",strlen(LAT_LONG_NAME),LAT_LONG_NAME);
  error(rc);
  rc=nc_put_att_text(ncid, lat_var_id,"standard_name",strlen(LAT_STANDARD_NAME),LAT_STANDARD_NAME);
  error(rc);
  rc=nc_put_att_text(ncid, lat_var_id,"axis",strlen("Y"),"Y");
  error(rc);
  dim[0]=lat_dim_id;
  dim[1]=lon_dim_id;
  rc=nc_def_var(ncid,"index",NC_INT,2,dim,&index_varid);
  error(rc);
  if(isnetcdf4 && compress)
  {
    rc=nc_def_var_deflate(ncid, index_varid, 0, 1,compress);
    error(rc);
  }
  rc=nc_put_att_text(ncid, index_varid,"standard_name",strlen("index"),"index");
  error(rc);
  rc=nc_put_att_text(ncid, index_varid,"long_name",strlen(INDEX_LONG_NAME),INDEX_LONG_NAME);
  error(rc);
  nc_put_att_int(ncid, index_varid,"missing_value",NC_INT,1,&miss);
  rc=nc_put_att_int(ncid, index_varid,"_FillValue",NC_INT,1,&miss);
  error(rc);
  if(header.nbands==2)
  {
    rc=nc_def_var(ncid,"riverlen",NC_FLOAT,2,dim,&len_varid);
    if(isnetcdf4 && compress)
    {
      rc=nc_def_var_deflate(ncid, len_varid, 0, 1,compress);
      error(rc);
    }
    rc=nc_put_att_text(ncid,len_varid,"units",strlen("m"),"m");
    error(rc);
    rc=nc_put_att_text(ncid, len_varid,"standard_name",strlen("river_length"),"river_length");
    error(rc);
    rc=nc_put_att_text(ncid, len_varid,"long_name",strlen("length of river"),"length of river");
    error(rc);
    fmiss=MISSING_VALUE_FLOAT;
    nc_put_att_float(ncid, len_varid,"missing_value",NC_FLOAT,1,&fmiss);
    rc=nc_put_att_float(ncid, len_varid,"_FillValue",NC_FLOAT,1,&fmiss);
    error(rc);
  }
  rc=nc_enddef(ncid);
  error(rc);
  rc=nc_put_var_double(ncid,lat_var_id,lat);
  error(rc);
  rc=nc_put_var_double(ncid,lon_var_id,lon);
  error(rc);
  out=newvec(int,nlon*nlat);
  check(out);
  for(i=0;i<nlon*nlat;i++)
    out[i]=miss;
  if(header.nbands==2)
  {
    len=newvec(float,nlon*nlat);
    check(len);
    for(i=0;i<nlon*nlat;i++)
      len[i]=fmiss;
  }
  for(i=0;i<n;i++)
  {
    src_cell=findcoord(grid+i,grid_soil,&resolution,n);
    if(src_cell==NOT_FOUND)
    {
      fprintf(stderr,"Source cell (%s) not found in `%s`.\n",
              sprintcoord(line,grid+i),argv[iarg]);
      return EXIT_FAILURE;
    }
    freadint(data,header.nbands,swap,file);
#ifdef DEBUG
    printf("data[0]=%d\n",data[0]);
#endif
    if(data[0]<0)
     out[index[src_cell]]=data[0];
    else
    {
      if(data[0]>=n)
      {
        fprintf(stderr,"Index for cell (%s)=%d in '%s' must be in <=%d.\n",
                sprintcoord(line,grid+i),data[0],argv[iarg+2],n-1);
        return EXIT_FAILURE;
      }
      dst_cell=findcoord(grid+data[0],grid_soil,&resolution,n);
      if(dst_cell==NOT_FOUND)
      {
        fprintf(stderr,"Destination cell (%s) not found in `%s`.\n",
                sprintcoord(line,grid+data[0]),argv[iarg]);
        return EXIT_FAILURE;
      }
      out[index[src_cell]]=index[dst_cell];
#ifdef DEBUG
      printf("out(%d]=%d\n",index[src_cell],index[dst_cell]);
#endif
    }
    if(header.nbands==2)
      len[index[src_cell]]=data[1];
  }
  rc=nc_put_var_int(ncid,index_varid,out);
  if(header.nbands==2)
    rc=nc_put_var_float(ncid,len_varid,len);
  nc_close(ncid);
  closecoord_netcdf(soil);
  return EXIT_SUCCESS;
#else
  fprintf(stderr,"ERROR401: NetCDF is not supported in this version of %s.\n",argv[0]);
  return EXIT_FAILURE;
#endif
} /* of 'main' */
